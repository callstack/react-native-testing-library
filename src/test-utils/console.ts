import { format } from 'util';

export function excludeConsoleMessage(logFn: (...args: unknown[]) => void, excludeMessage: string) {
  return (...args: unknown[]) => {
    const message = format(...args);
    if (message.includes(excludeMessage)) {
      return;
    }

    logFn(...args);
  };
}
