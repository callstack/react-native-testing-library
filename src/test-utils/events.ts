interface EventEntry {
  name: string;
  payload: any;
}

export function createEventLogger() {
  const events: EventEntry[] = [];
  const logEvent = (name: string) => {
    return (event: unknown) => {
      const eventEntry: EventEntry = {
        name,
        payload: event,
      };

      events.push(eventEntry);
    };
  };

  return { events, logEvent };
}
