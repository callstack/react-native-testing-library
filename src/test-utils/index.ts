export * from './console';
export * from './events';
