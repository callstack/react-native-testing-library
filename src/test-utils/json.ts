import type { ReactTestRendererJSON } from 'react-test-renderer';

type JsonPropsMapper = {
  [key: string]: unknown;
};

export function mapJsonProps<T extends ReactTestRendererJSON | ReactTestRendererJSON[] | null>(
  element: T,
  mapper: JsonPropsMapper,
): T {
  if (Array.isArray(element)) {
    return element.map((e) => mapJsonProps(e, mapper)) as T;
  }

  if (!element) {
    return element;
  }

  const resultProps = { ...element.props };
  Object.keys(mapper).forEach((key) => {
    if (key in element.props) {
      resultProps[key] = mapper[key];
    }
  });

  const resultElement = { ...element, props: resultProps };
  Object.defineProperty(resultElement, '$$typeof', {
    value: Symbol.for('react.test.json'),
  });

  return resultElement;
}
