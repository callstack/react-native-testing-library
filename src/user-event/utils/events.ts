import { ReactTestInstance } from 'react-test-renderer';
import act from '../../act';
import { isEventEnabled, isTouchResponder } from '../../fireEvent';

type EventHandler = (event: unknown) => void;

/**
 * Dispatch event function used by User Event module.
 *
 * @param element element trigger event on
 * @param eventName name of the event
 * @param event event payload
 */
export function dispatchHostEvent(
  element: ReactTestInstance,
  eventName: string,
  event: unknown
) {
  const handler = getEnabledEventHandler(element, eventName);
  if (!handler) {
    return;
  }

  act(() => {
    handler(event);
  });
}

function getEnabledEventHandler(
  element: ReactTestInstance,
  eventName: string
): EventHandler | null {
  const touchResponder = isTouchResponder(element) ? element : undefined;

  const handler = getEventHandler(element, eventName);
  if (handler && isEventEnabled(element, eventName, touchResponder)) {
    return handler;
  }

  return null;
}

function getEventHandler(element: ReactTestInstance, eventName: string) {
  const eventHandlerName = getEventHandlerName(eventName);
  if (typeof element.props[eventHandlerName] === 'function') {
    return element.props[eventHandlerName];
  }

  return undefined;
}

function getEventHandlerName(eventName: string) {
  return `on${eventName.charAt(0).toUpperCase()}${eventName.slice(1)}`;
}
