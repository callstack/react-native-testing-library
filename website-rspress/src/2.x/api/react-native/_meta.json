[
  {
    "type": "file",
    "name": "index",
    "label": "Table of Contents"
  },
  {
    "type": "dir",
    "name": "classes",
    "label": "Classes"
  },
  {
    "type": "dir",
    "name": "interfaces",
    "label": "Interfaces"
  },
  {
    "type": "dir",
    "name": "types",
    "label": "Types"
  }
]
