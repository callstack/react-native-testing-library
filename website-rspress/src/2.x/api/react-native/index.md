# Re.Pack React Native APIs

## Table of contents

### Classes

- [Chunk](./classes/Chunk.md)
- [ChunkManager](./classes/ChunkManager.md)

### Interfaces

- [ChunkConfig](./interfaces/ChunkConfig.md)
- [ChunkManagerConfig](./interfaces/ChunkManagerConfig.md)
- [RemoteChunkLocation](./interfaces/RemoteChunkLocation.md)
- [StorageApi](./interfaces/StorageApi.md)

### Type aliases

- [RemoteChunkResolver](./types/RemoteChunkResolver.md)
