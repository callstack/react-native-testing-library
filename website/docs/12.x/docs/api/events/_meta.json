[
  { "type": "file", "name": "user-event", "label": "User Event" },
  { "type": "file", "name": "fire-event", "label": "Fire Event" }
]
