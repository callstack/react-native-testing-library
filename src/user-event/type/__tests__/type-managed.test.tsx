import * as React from 'react';
import { TextInput } from 'react-native';
import { createEventLogger } from '../../../test-utils/events';
import { render } from '../../..';
import { userEvent } from '../..';

beforeEach(() => {
  jest.useRealTimers();
});

interface ManagedTextInputProps {
  valueTransformer?: (text: string) => string;
  logEvent: (name: string) => (event: any) => void;
  initialValue?: string;
}

function ManagedTextInput({
  logEvent,
  valueTransformer,
  initialValue = '',
}: ManagedTextInputProps) {
  const [value, setValue] = React.useState(initialValue);

  const handleChangeText = (text: string) => {
    logEvent('changeText')(text);
    const newValue = valueTransformer?.(text) ?? text;
    setValue(newValue);
  };

  return (
    <TextInput
      testID="input"
      value={value}
      onChangeText={handleChangeText}
      onFocus={logEvent('focus')}
      onBlur={logEvent('blur')}
      onPressIn={logEvent('pressIn')}
      onPressOut={logEvent('pressOut')}
      onChange={logEvent('change')}
      onKeyPress={logEvent('keyPress')}
      onTextInput={logEvent('textInput')}
      onSelectionChange={logEvent('selectionChange')}
      onSubmitEditing={logEvent('submitEditing')}
      onEndEditing={logEvent('endEditing')}
      onContentSizeChange={logEvent('contentSizeChange')}
    />
  );
}

describe('type() for managed TextInput', () => {
  it('supports basic case', async () => {
    jest.spyOn(Date, 'now').mockImplementation(() => 100100100100);
    const { events, logEvent } = createEventLogger();
    const screen = render(<ManagedTextInput logEvent={logEvent} />);

    const user = userEvent.setup();
    await user.type(screen.getByTestId('input'), 'Wow');

    const eventNames = events.map((e) => e.name);
    expect(eventNames).toEqual([
      'pressIn',
      'focus',
      'pressOut',
      'keyPress',
      'change',
      'changeText',
      'selectionChange',
      'keyPress',
      'change',
      'changeText',
      'selectionChange',
      'keyPress',
      'change',
      'changeText',
      'selectionChange',
      'endEditing',
      'blur',
    ]);

    expect(events).toMatchSnapshot('input: "Wow"');
  });

  test('supports rejecting TextInput', async () => {
    jest.spyOn(Date, 'now').mockImplementation(() => 100100100100);
    const { events, logEvent } = createEventLogger();
    const screen = render(
      <ManagedTextInput
        initialValue="XXX"
        logEvent={logEvent}
        valueTransformer={() => 'XXX'}
      />
    );

    const user = userEvent.setup();
    await user.type(screen.getByTestId('input'), 'ABC');

    const eventNames = events.map((e) => e.name);
    expect(eventNames).toEqual([
      'pressIn',
      'focus',
      'pressOut',
      'keyPress',
      'change',
      'changeText',
      'selectionChange',
      'keyPress',
      'change',
      'changeText',
      'selectionChange',
      'keyPress',
      'change',
      'changeText',
      'selectionChange',
      'endEditing',
      'blur',
    ]);

    expect(events).toMatchSnapshot('input: "ABC", value: "XXX"');
  });
});
