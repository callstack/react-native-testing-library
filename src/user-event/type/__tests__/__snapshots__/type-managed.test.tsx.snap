// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`type() for managed TextInput supports basic case: input: "Wow" 1`] = `
[
  {
    "name": "pressIn",
    "payload": {
      "currentTarget": {
        "measure": [Function],
      },
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "changedTouches": [],
        "identifier": 0,
        "locationX": 0,
        "locationY": 0,
        "pageX": 0,
        "pageY": 0,
        "target": 0,
        "timestamp": 100100100100,
        "touches": [],
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "focus",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "target": 0,
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "pressOut",
    "payload": {
      "currentTarget": {
        "measure": [Function],
      },
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "changedTouches": [],
        "identifier": 0,
        "locationX": 0,
        "locationY": 0,
        "pageX": 0,
        "pageY": 0,
        "target": 0,
        "timestamp": 100100100100,
        "touches": [],
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "keyPress",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "key": "W",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "change",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "eventCount": 0,
        "target": 0,
        "text": "W",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "changeText",
    "payload": "W",
  },
  {
    "name": "selectionChange",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "selection": {
          "end": 1,
          "start": 1,
        },
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "keyPress",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "key": "o",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "change",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "eventCount": 0,
        "target": 0,
        "text": "Wo",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "changeText",
    "payload": "Wo",
  },
  {
    "name": "selectionChange",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "selection": {
          "end": 2,
          "start": 2,
        },
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "keyPress",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "key": "w",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "change",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "eventCount": 0,
        "target": 0,
        "text": "Wow",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "changeText",
    "payload": "Wow",
  },
  {
    "name": "selectionChange",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "selection": {
          "end": 3,
          "start": 3,
        },
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "endEditing",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "target": 0,
        "text": "Wow",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "blur",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "target": 0,
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
]
`;

exports[`type() for managed TextInput supports rejecting TextInput: input: "ABC", value: "XXX" 1`] = `
[
  {
    "name": "pressIn",
    "payload": {
      "currentTarget": {
        "measure": [Function],
      },
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "changedTouches": [],
        "identifier": 0,
        "locationX": 0,
        "locationY": 0,
        "pageX": 0,
        "pageY": 0,
        "target": 0,
        "timestamp": 100100100100,
        "touches": [],
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "focus",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "target": 0,
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "pressOut",
    "payload": {
      "currentTarget": {
        "measure": [Function],
      },
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "changedTouches": [],
        "identifier": 0,
        "locationX": 0,
        "locationY": 0,
        "pageX": 0,
        "pageY": 0,
        "target": 0,
        "timestamp": 100100100100,
        "touches": [],
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "keyPress",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "key": "A",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "change",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "eventCount": 0,
        "target": 0,
        "text": "XXXA",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "changeText",
    "payload": "XXXA",
  },
  {
    "name": "selectionChange",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "selection": {
          "end": 4,
          "start": 4,
        },
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "keyPress",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "key": "B",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "change",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "eventCount": 0,
        "target": 0,
        "text": "XXXB",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "changeText",
    "payload": "XXXB",
  },
  {
    "name": "selectionChange",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "selection": {
          "end": 4,
          "start": 4,
        },
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "keyPress",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "key": "C",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "change",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "eventCount": 0,
        "target": 0,
        "text": "XXXC",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "changeText",
    "payload": "XXXC",
  },
  {
    "name": "selectionChange",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "selection": {
          "end": 4,
          "start": 4,
        },
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "endEditing",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "target": 0,
        "text": "XXX",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "blur",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "target": 0,
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
]
`;
